src/konserve/filestore.clj

```diff
@@ -49,454 +45,439 @@
+                                                     :folder path})))
+        (.delete test-file)))
+
+(defn filestore-schema-update
+  "Lists all keys in this binary store. This operation *does not block concurrent operations* and might return an outdated key set. Keys of binary blobs are not tracked atm."
+  [{:keys [folder serializer read-handlers] :as store}]
+  (let [fns (->> (io/file folder)
+                 .list
+                 seq
+                 ((fn [e] (prn "foo: " e) e))
+                 (filter #(re-matches #"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
+                                      %))
+                 (map (fn [fn]
+                        (go-locked
+                         store fn
+                         (let [f (io/file (str folder "/" fn))]
+                           (when (.exists f)
+                             (let [fis (DataInputStream. (FileInputStream. f))]
+                               (try
+                                 (let [[key value] (-deserialize serializer read-handlers fis)]
+                                   (<! (-assoc-in store key value))
+                                   (.delete f))
+                                 (catch Exception e
+                                   (ex-info "Could not read key."
+                                            {:type :read-error
+                                             :key fn
+                                             :exception e}))
+                                 (finally
+                                   (.close fis)))))))))
+                 async/merge
+                 (async/into #{}))]
+        fns))
 
+(defn new-fs-store
+  "Filestore contains a Key and a Data Folder"
+  [path  & {:keys [serializer read-handlers write-handlers config automatic-schema-update]
+            :or   {serializer     (ser/fressian-serializer)
+                   read-handlers  (atom {})
+                   write-handlers (atom {})
+                   config         {:fsync true}
+                   automatic-schema-update false}}]
+  (println "auto: " automatic-schema-update)
+  (let [_          (check-and-create-folder path)
+        old-format? (not (empty? (filter #(re-matches #"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}" %)
+                                                      (seq (.list (io/file path))))))
+        key-path   (str path "/Key")
+        data-path  (str path "/Data")
+        _          (check-and-create-folder key-path)
+        _          (check-and-create-folder data-path)
+        store (map->FileSystemStore {:folder         path
+                                     :serializer     serializer
+                                     :read-handlers  read-handlers
+                                     :write-handlers write-handlers
+                                     :locks          (atom {})})]
+    (if (and old-format? (true? automatic-schema-update))
```

Comment:
We forgot to throw in the case where `(and old-format? (not automatic-schema-update))`. true? is not necessary here btw.